### PR TITLE
Add MFTLoss class for Minifinetuning self-corrective distillation implementation

### PR DESCRIPTION
**Pull request by pbelcak**
Add MFTLoss class for KL-divergence with Minifinetuning threshold modification. MFT was proposed by NVR in the following tech report: https://arxiv.org/pdf/2506.15702.

## What does this PR do?

**Type of change:** New feature

**Overview:** This PR adds a new `MFTLoss` class that implements KL-divergence loss with Minifinetuning threshold modification as described in the [paper](https://arxiv.org/abs/2506.15702). The implementation includes a sophisticated distribution correction mechanism that adjusts teacher distributions based on ground truth labels and a configurable threshold parameter.

## Usage

```python
import torch
from modelopt.torch.distill.losses import MFTLoss

# Initialize the MFT loss
mft_loss = MFTLoss(temperature=1.0, reduction="batchmean")

# Example usage with student logits, teacher logits, labels, and threshold
logits_s = torch.randn(2, 10, 1000)  # Student logits (batch_size, seq_len, vocab_size)
logits_t = torch.randn(2, 10, 1000)  # Teacher logits (batch_size, seq_len, vocab_size)
labels = torch.randint(0, 1000, (2, 10))  # Ground truth labels (batch_size, seq_len)
threshold = 0.2

# Compute the MFT loss
loss = mft_loss(logits_s, logits_t, labels, threshold)
```

## Testing
Extensive functional testing was conducted for the paper. Tested the implementation for ModelOpt manually with toy examples such as the usage example above.

## Before your PR is "*Ready for review*"

- **Make sure you read and follow [Contributor guidelines](https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/main/CONTRIBUTING.md)**
- **Is this change backward compatible?**: Yes.
- **Did you write any new necessary tests?**: Yes, under `tests/unit/torch/distill/test_distill.py::test_distillation_model_mft`.
- **Did you add or update any necessary documentation?**: Yes, under `docs/source/guides/4_distillation.rst`.
- **Did you update [Changelog](https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/main/CHANGELOG.rst)?**: Yes.

### CODE REVIEW

**Comment by kevalmorabia97** [comment]
Closing as this will be merged separately
